crates/biome_deserialize/src/json.rs

```diff
@@ -245,15 +245,28 @@ impl DeserializableValue for JsonMemberName {
         name: &str,
         diagnostics: &mut Vec<DeserializationDiagnostic>,
     ) -> Option<V::Output> {
-        let value = self.inner_string_text().ok()?;
-        visitor.visit_str(Text(value), AstNode::range(self), name, diagnostics)
+        let value = unescape_json(self.inner_string_text().ok()?);
+        visitor.visit_str(value, AstNode::range(self), name, diagnostics)
```

Comment:
Escaping/Unescaping is always mind-bending to me. Could you please elaborate why we want to do this? Like, what was the issue and what difference does it make?

Reply:
`useNamingConvention` accepts an option `conventions` that allows adding custom conventions.
A custom convention may express a constraint as a regex.
For instance, you can require that every variable ends with `$`.
In a regex you have to escape `$`, otherwise it is an anchor.
JSON doesn't allow unescaped backsash `\`. Thus you have to write `\\$`.
Unfortunately, our desewrailizer doesn't turn `\\` into `\`.
Thus, we end up with `\` being escaped, not `$`.
We should turn `\\$` into `\$` during the deserialization.
This is the reason d'être of this fix.

Reply:
Thank you for such detailed explanation!

But I still have a question, does this fix mean that for the following `JSON` object:

```json
{
  "a": "\\$"
}
```

The sytax tree will have this change? 🤔

```diff
value: JsonStringValue {
-          value_token: JSON_STRING_LITERAL@9..14 "\"\\\\$\"" [] [],
+          value_token: JSON_STRING_LITERAL@9..14 "\"\\$\"" [] [],
        },
```

Reply:
Oh, I just realized this is only for biome deserialization, so the syntax tree is not at play here. I think I misread the code, sorry.

Reply:
Yeah, this only affects `biome_deserialize`.

Perhaps another issue we have is when we create new JSON string nodes: currently we didn't escape the string. This can lead to a parsing error. We could fix it in another PR.